### PR TITLE
[chore] enable coverage for unit tests

### LINKED ISSUE #8394: Reenable codecov integration
Lately I've seen a couple of bug fixes of obvious code issues like #8392.

A while ago codecov was disabled in this repository in light of its security incident but I think it is appropriate to reenable it as the CI has been migrated to GHA which has a lower vulnerability surface than Circle did (the step will only have secrets passed explicitly). Or even more strict would be to only enable for PRs but not for main branch to ensure 100% that no secrets can be leaked. It's currently very hard to ensure good code coverage when reviewing PRs, nor should it be a manual process like that but should be assisted by a tool like codecov.

### PR DESCRIPTION
**Pull request by codeboten**
This re-enables unit test coverage for the repository. Coverage will not immediately include integration tests.

Fixes #8394

### CODE REVIEW

**Review by dmitryax — APPROVED** [review]

**Review by MovieStoreGuy — APPROVED** [review]
Thank you for adding this, I appreciate being able to run CI steps locally using make :D

**Review by atoulme — APPROVED** [review]

**Comment by codeboten** [comment]
Note that I ran into [this](https://github.com/codecov/codecov-action/issues/926) issue when enabling coverage. We may want to disable ci failures for coverage uploads. Will need to keep an eye on how frequently this impacts the builds.

File: Makefile.Common [hunk 1]
```diff
@@ -96,7 +96,12 @@ common: lint
 .PHONY: test
 test:
 	$(GOTEST) $(GOTEST_OPT) ./...
-	
+
+.PHONY: test-with-cover
```
[the remark this answers is not in the document]
Comment (codeboten) [reply Makefile.Common:100]: Yeah that's a follow up PR, i wanted to keep it until i figured out the integration tests :)

[the remark this answers is not in the document]
Comment (djaglowski) [reply Makefile.Common:100]: Possibly relevant, in case you haven't seen it: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/22808

**Pull request merged**